### PR TITLE
Introduce concept of floundering

### PR DESCRIPTION
**Pull request by nikomatsakis**
This introduces the idea of "floundering" into chalk-engine and uses it to mark auto-traits as non-enumerable, along with a new `#[non_enumerable]` traits (used for testing purposes). See the edits to the README.md in one of the final commits for a high-level summary.

Fixes https://github.com/rust-lang/chalk/issues/217

### CODE REVIEW

**Comment by shepmaster** [comment]
I’d like to requisition the addition of the turboflounder.

**Comment by nikomatsakis** [comment]
(The error on stable seems to have something to do with generating documentation, I think -- the tests run successfully.)

**Comment by nikomatsakis** [comment]
I was thinking about this some more. We'll see, but I imagine we'll want to tinker the story a bit to try and give better "type hints" in some cases. I've not yet come up with a complete example but the idea is something like:

```rust
trait Foo { }
impl<T> Foo for Vec<T> { }
```

Here, if yo have `?T: Foo`, it's going to flounder, but it could in principle infer that `?T = Vec<?U>` for *some* `?U`.

**Comment by matklad** [comment]
@nikomatsakis just out of curiosity, could you briefly explain (or point to an existing explanation) *why* the need to enumerate solutions arises at all? From my little understanding of how the trait solving works, it seems like one needs only to check things, and not to enumerate them. From your comment I gather that this can be useful at least for error reporting, but I wonder if, when doing completions, we can just disable error reporting? Or is there a more fundamental reason why we need to enumerate things?

File: chalk-engine/src/README.md [hunk 1]
```diff
@@ -231,6 +240,122 @@ Here you can see how the forest captures both the answers we have
 created thus far *and* the strands that will let us try to produce
 more answers later on.
 
+### Floundering
+
+The first thing we do when we create a table is to initialize it with
+a set of strands. These strands represent all the ways that one can
+solve the table's associated goal. For an ordinary trait, we would
+effectively create one strand per "relevant impl". But sometimes the
+goals are vague for this to be possible; other times, it may be possible
```
Comment (tmandry) [inline chalk-engine/src/README.md]: nit: "_too_ vague"

Comment (tmandry) [reply chalk-engine/src/README.md]:
```suggestion
goals are too vague for this to be possible; other times, it may be possible
```

**Comment by scalexm** [comment]
@matklad I believe you'd need to enumerate things when inferring types from some trait method calls, e.g. like in:
```rust
struct Bar<T>(T);

trait Foo {
    fn foo(&self);
}

impl Foo for Bar<u64> {
    fn foo(&self) { }
}

impl Bar<u64> {
    fn print_u64(&self) {
        println!("hello u64");
    }
}

impl Bar<u8> {
    fn print_u8(&self) {
        println!("hello u8");
    }
}

fn main() {
    let x = Bar(5);

    // To infer the type of `x: Bar<?X>`, we'd need to solve the following goal:
    // `exists<?X> { Implemented(Bar<?X>: Foo) }`
    // In this case, the goal won't flounder and `?X` will resolve to `u64`.
    x.foo();

    // auto-completion: `print_` -> `print_u64()`, because `x` has been inferred
    // to be of type `Bar<u64>`.
    y.print_
}
```

If the goal had returned more than one answer or had floundered, you'd report an error.

**Pull request merged**